src/network/NetworkManagerPrimary.cc

```diff
@@ -129,55 +133,75 @@ bool NetworkManagerPrimary::Step(const UpdateInfo &_info)
     return false;
   }
 
-  private_msgs::SimulationStep step;
-  step.mutable_stats()->CopyFrom(convert<msgs::WorldStatistics>(_info));
-
-  // Affinities that changed this step
-  this->PopulateAffinities(step);
-
-  // Check all secondaries are ready to receive steps - only do this once at
-  // startup
-  if (!this->SecondariesCanStep())
-  {
-    return false;
+  // TODO(ivanpauno): If secondaries received a step message allowing them to move ahead N iterations (hardcoded to 1000 now),
+  // until the secondaries completed those N steps the simulation cannot be paused, the step size cannot be changed, etc.
+  // This should be handle in a better fashion.
+  // Note: send an ack each N/2 iterations, to allow secondaries to move ahead faster.
+  if (((0uLL == _info.iterations % (kSecondaryIterations/2)) || this->paused) && !_info.paused) {
+    // Allow secondaries to continue moving forward each N steps (1000).
+    // Also send a message if the simulation was paused before and now is running.
+    private_msgs::SimulationStep step;
+    step.mutable_stats()->CopyFrom(convert<msgs::WorldStatistics>(_info));
+    step.set_max_iterations(_info.iterations + kSecondaryIterations);
+
+    // TODO(ivanpauno): Affinities should only be calculated at startup.
+    // Then we should have logic to detect if performers are "far apart", "viewing each other", "interacting".
+    // In the first case secondaries can run asynchronously (implemented).
+    // In the second a perfect lockstep is needed (TODO).
+    // In the third case, the physics should be simulated in the same secondary,
+    // and we need to run the preUpdate/Update/postUpdate in a perfect lockstep fashion.
+    this->PopulateAffinities(step);
+    this->simStepPub.Publish(step);
   }
+  this->paused = _info.paused;
 
-  // Send step to all secondaries
-  this->secondaryStates.clear();
-  this->secondaryStatesPromise = std::promise<void>{};
-  auto future = this->secondaryStatesPromise.get_future();
-  this->simStepPub.Publish(step);
+  if (_info.iterations >= this->nextIteration && !_info.paused) {
+    // Update state based on secondaries messages.
+    std::vector<private_msgs::SecondaryStep> secondariesSteps;
+    {
+      std::unique_lock<std::mutex> guard{this->secondaryStatesMutex};
 
-  // Block until all secondaries are done
```

Comment:
Maybe it's good to keep this comment, as the next `for loop` is doing that

Reply:
Agreed, let's restore that.